client/mm/libxc/binance.go

```diff
@@ -47,38 +47,180 @@ const (
+			}
+
+			convertedUpdates = append(convertedUpdates, &obEntry{
+				rate: calc.MessageRateAlt(price, b.baseConversionFactor, b.quoteConversionFactor),
+				qty:  uint64(qty * float64(b.baseConversionFactor)),
+			})
+		}
+
+		return convertedUpdates, nil
+	}
+
+	bids, err = convert(binanceBids)
+	if err != nil {
+		return nil, nil, err
 	}
+
+	asks, err = convert(binanceAsks)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return bids, asks, nil
 }
 
-type bncAssetConfig struct {
-	// symbol is the DEX asset symbol, always lower case
-	symbol string
-	// coin is the asset symbol on binance, always upper case.
-	// For a token like USDC, the coin field will be USDC, but
-	// symbol field will be usdc.eth.
-	coin string
-	// chain will be the same as coin for the base assets of
-	// a blockchain, but for tokens it will be the chain
-	// that the token is hosted such as "ETH".
-	chain            string
-	conversionFactor uint64
+// sync does an initial sync of the orderbook. When the first update is
+// received, it grabs a snapshot of the orderbook, and only processes updates
+// that come after the state of the snapshot. A goroutine is started that keeps
+// that keeps the orderbook in sync by repeating the sync process if an update
+// is ever missed.
+//
+// This function runs until the context is canceled. It must be started as
+// a new goroutine.
+func (b *binanceOrderBook) sync(ctx context.Context, getSnapshot func() (*binanceOrderbookSnapshot, error)) {
+	syncOrderbook := func(minUpdateID uint64) (uint64, bool) {
+		b.log.Debugf("Syncing %s orderbook. First update ID: %d", b.mktID, minUpdateID)
+
+		const maxTries = 5
+		for i := 0; i < maxTries; i++ {
+			if i > 0 {
+				time.Sleep(2 * time.Second)
+			}
+
+			snapshot, err := getSnapshot()
+			if err != nil {
+				b.log.Errorf("Error getting orderbook snapshot: %v", err)
+				time.Sleep(time.Second)
+				continue
+			}
+
+			if snapshot.LastUpdateID >= minUpdateID {
+				bids, asks, err := b.convertBinanceBook(snapshot.Bids, snapshot.Asks)
+				if err != nil {
+					b.log.Errorf("Error parsing binance book: %v", err)
+					return 0, false
+				}
+
+				b.log.Debugf("Got %s orderbook snapshot with update ID %d", b.mktID, snapshot.LastUpdateID)
+
+				b.book.clear()
+				b.book.update(bids, asks)
+				return snapshot.LastUpdateID, true
+			} else {
+				b.log.Infof("Snapshot last update ID %d is less than first update ID %d. Getting new snapshot...", snapshot.LastUpdateID, minUpdateID)
+			}
+		}
+
+		return 0, false
+	}
+
+	setSynced := func(synced bool) {
+		b.mtx.Lock()
+		b.synced = synced
+		b.mtx.Unlock()
+	}
+
+	firstUpdate := <-b.updateQueue
+	latestUpdate, success := syncOrderbook(firstUpdate.FirstUpdateID)
+	if !success {
+		b.log.Errorf("Failed to sync %s orderbook", b.mktID)
+		return
+	}
+
+	setSynced(true)
+
+	for {
+		select {
+		case update := <-b.updateQueue:
+			if update.LastUpdateID <= latestUpdate {
+				continue
+			}
+
```

Comment:
`setSynced(false)` doesn't need to be called here. This is the case where the local order book is already ahead of the state in the update. This will usually be hit on the initial sync if more than one update was received before getting the snapshot.